### ISSUE TITLE
Staleness_interval for Streaming aggregation does not work corect?

### ISSUE THREAD

**Issue by mrsanuk** [VictoriaMetrics/VictoriaMetrics#7116]
### Is your question request related to a specific component?

vmagent

### Describe the question in detail

We have few types of job that generate metrics push it to vmagent with streaming aggregation and vmagent writes to the prometheus, one runs every 10 min, second once a hour. To avoid no data in prometheus we setted up staleness_interval. 
Our config:
```
      streamaggr.yml: |
        - match: 
          - report_counter
          interval: 2m
          outputs: ["total"]
          without: ["job"]
          staleness_interval: 70m
          keep_metric_names: true
          flush_on_shutdown: true
```
Problem only that first metric after vmgent start I can see only after staleness_interval, if it 70 min, 70 min will be no data in prometheus, if 15 min then 15 min no data,  if 1 day then 1 day nodata. After that everything works fine. Is it a bug or expected behavior?

### Troubleshooting docs

- [ ] General - https://docs.victoriametrics.com/troubleshooting/
- [X] vmagent - https://docs.victoriametrics.com/vmagent/#troubleshooting
- [ ] vmalert - https://docs.victoriametrics.com/vmalert/#troubleshooting

**Comment by AndrewChubatiuk** [comment]
hey @mrsanuk
suppose we need to add additional variable as staleness interval now is used to determine:
- how much time state is for each output will be stored
- how much [time to wait](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/f65b976eda7f7e29010f35ad54e89da421cfa7aa/lib/streamaggr/total.go#L63) since vmagent start before allowing to store first samples
@hagen1778 what do you think about removing second scenario completely since we have `ignore_first_samples`?

**Comment by hagen1778** [comment]
> how much [time to wait](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/f65b976eda7f7e29010f35ad54e89da421cfa7aa/lib/streamaggr/total.go#L63) since vmagent start before allowing to store first samples
@hagen1778 what do you think about removing second scenario completely since we have ignore_first_samples?

@AndrewChubatiuk I'm not sure. Is this the only reason why https://github.com/VictoriaMetrics/VictoriaMetrics/commit/138a4d1c2badb59b284e994fd017d6f8ab3c368c was added?

**Comment by AndrewChubatiuk** [comment]
I don't see any other explanation for this functionality besides avoid having spikes during startup

**Comment by iyuroch** [comment]
@AndrewChubatiuk @hagen1778 Reading this PR I'm still a bit at a loss as to why this was added, I understand that it can add "spike" at start but what difference does it make if "spike" is now or after some time? Cardinality will be the same or even larger at the end of the staleness interval.

**Comment by iyuroch** [comment]
And to give more context - we use vmagent as pushgateway replacement for job metrics, some jobs we run couple of times a day, and with low staleness interval we are missing metrics aggregation, at high staleness intervals - they don't appear.

**Comment by oveleten** [comment]
@AndrewChubatiuk so, are you going to provide some fix in the next versions?

**Comment by iyuroch** [comment]
@AndrewChubatiuk @hagen1778 I've opened the following PR with a proposed separate flag that adds the ability to control this behavior:
https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7313
Let me know what you think about it.

**Comment by fxrlv** [comment]
> @AndrewChubatiuk @hagen1778 Reading this PR I'm still a bit at a loss as to why this was added, I understand that it can add "spike" at start but what difference does it make if "spike" is now or after some time? Cardinality will be the same or even larger at the end of the staleness interval.

I have tested streaming aggregation for a while and can assume it was added for the next reason:
Imagine having `total` streamaggr output for a constant counter, for instance `cpu_limit_total` over `cpu_limit`. Constant here is just for simplicity.

Once vmagent restarted it will receive `cpu_limit` samples, say 100. vmagent hasn't seen this timeseries yet, so it has two options: 
1) increase output `total` to 100 (treat it like a counter reset) and set `lastSeenValue` (internal state) to 100
2) keep `total` 0 and set `lastSeenValue` to 100

the first option doesn't work fine:
say we want to apply `increase` to `cpu_limit_total`, `increase` assumes counters start from 0
that way `increase(cpu_limit_total)` will produce a huge spike since the value _is increased from 0_ to 100

the second option is what `IgnoreFirstSampleInterval` does
vmagent observes the counter and its current value but doesn't increase `total` output since **it can't be sure whether it's a new timeseries or not, it just takes some time to warmup**

**Comment by iyuroch** [comment]
@fxrlv thanks for this example, so this specific case will track it for longer, and because we save the last value we will add a difference between the current state and the last value thus eliminating spike. Is this the correct statement?
I do see a bit of a problem here - let's say it's the real first value, in that case, I'll miss this initial spike.
Another one - is that when the application is restarted during that time we will not have a sum of the total and the counter value before restart will not be accounted. I don't know if I understood these implications if you can comment on them?

Also, the proposed code changes to this logic we are discussing are here: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7313. Could you also check them from your side, as they include test case changes, and you seem to have knowledge of expected behavior?

**Comment by fxrlv** [comment]
> because we save the last value we will add a difference between the current state and the last value thus eliminating spike. Is this the correct statement?

I don't think so. Consider constant samples with the given timeline:
1) -> sample=100
2) -> vmagent gets restarted (`lastSeenValue=0` `total=0`)
3) -> sample=100 (set `lastSeenValue=100`, do not change `total` due to `ignoreFirstSampleInterval` so `total=0`)
4) -> sample=100 (set `lastSeenValue=100`, `total=value-lastSeenValue` so `total=0`)

As shown here there is no spike once we've reached `ignoreFirstSampleDeadline`

> I do see a bit of a problem here - let's say it's the real first value, in that case, I'll miss this initial spike.

That's what the comment below tries to clarify saying `at startup when new time series cannot be distinguished from already existing series`.

https://github.com/VictoriaMetrics/VictoriaMetrics/blob/5d73b8b86646a5ded5a6880308b9f5893b133854/lib/streamaggr/total.go#L28-L31

There is no way to distinguish those two cases. Since restarting vmagent already produce a gap (missing samples), what's the reason of preferring a spike here?

> Another one - is that when the application is restarted during that time we will not have a sum of the total and the counter value before restart will not be accounted.

It shouldn't since streamaggr's `total` is not PromQL's `sum`. Maybe you need [sum_samples](https://docs.victoriametrics.com/stream-aggregation/#sum_samples).

**Comment by fxrlv** [comment]
`total_prometheus` doesn't count all first samples of new timeseries vmagent hasn't seen before, not only the first ones after vmagent's restart

**Comment by AndrewChubatiuk** [comment]
okay, so in this case there's `ignore_first_intervals` parameter, which skips flushing of first n intervals after vmagent start

**Comment by fxrlv** [comment]
how does it help?

1) -> sample=100
2) -> vmagent gets restarted (`lastSeenValue=0` `total=0`)
3) -> sample=100 (set `lastSeenValue=100`, `total=100`, do not flush since `ignore_first_intervals=1`)
4) -> sample=100 (set `lastSeenValue=100`, `total+=value-lastSeenValue` so `total=100`, do flush)

resulting to a counter with initial value 100 which causes a spike

**Comment by iyuroch** [comment]
@fxrlv, while I see what problem you are describing, however, current spike protection, causes another one which we have:
1. data point 100 0
2. data point 150 1
3. vmagent shows 50, should be 150 instead

**Comment by iyuroch** [comment]
in this case - yes, we have many short lived jobs that process tasks, when job start - it has spike in metric and then may add there a bit. right now during vmagent start we are missing this data of initial processed tasks. moreover, due to staleness interval == for how much we are skipping metrics during vmagent restart we do not have data for quite some time (we want staleness interval to be > 2h but we can't as during restart we will miss this data).

**Comment by fxrlv** [comment]
I see your point, thanks!

Then I think your initial idea of https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7313 having dedicated knob called `IgnoreFirstSampleInterval` would solve all the issues

**Comment by f41gh7** [comment]
The feature was included into [v1.107.0](https://docs.victoriametrics.com/changelog/#v11070)
